File: ccl/platform/android/vulkan/vulkansurfaceview.cpp
//************************************************************************************************
//
// This file is part of Crystal Class Library (R)
// Copyright (c) 2025 CCL Software Licensing GmbH.
// All Rights Reserved.
//
// Licensed for use under either:
//  1. a Commercial License provided by CCL Software Licensing GmbH, or
//  2. GNU Affero General Public License v3.0 (AGPLv3).
// 
// You must choose and comply with one of the above licensing options.
// For more information, please visit ccl.dev.
//
// Filename    : ccl/platform/android/vulkan/vulkansurfaceview.cpp
// Description : Vulkan surface view JNI interface
//
//************************************************************************************************

#include "ccl/platform/android/vulkan/vulkansurfaceview.h"
#include "ccl/platform/android/vulkan/vulkanrendertarget.android.h"

#include "ccl/platform/android/gui/androidview.h"
#include "ccl/platform/android/gui/frameworkactivity.h"
#include "ccl/platform/android/gui/frameworkview.h"
#include "ccl/platform/android/gui/window.android.h"

#include "ccl/base/message.h"

#include "ccl/gui/gui.h"

#include <android/native_window_jni.h>

//************************************************************************************************
// JNI classes
//************************************************************************************************

namespace CCL {
namespace Android {

//************************************************************************************************
// RenderSurfaceView
//************************************************************************************************

DECLARE_JNI_CLASS (RenderSurfaceView, CCLGUI_CLASS_PREFIX "VulkanSurfaceView")
	DECLARE_JNI_CONSTRUCTOR (construct, jobject)
	DECLARE_JNI_METHOD (void, layout, int, int, int, int)
	DECLARE_JNI_METHOD (void, startRendering)
	DECLARE_JNI_METHOD (void, stopRendering)
	DECLARE_JNI_METHOD (bool, isAlive)
END_DECLARE_JNI_CLASS (RenderSurfaceView)

DEFINE_JNI_CLASS (RenderSurfaceView)
	DEFINE_JNI_CONSTRUCTOR (construct, "(JLandroid/content/Context;)V")
	DEFINE_JNI_METHOD (layout, "(IIII)V")
	DEFINE_JNI_METHOD (startRendering, "()V")
	DEFINE_JNI_METHOD (stopRendering, "()V")
	DEFINE_JNI_METHOD (isAlive, "()Z")
END_DEFINE_JNI_CLASS

} // namespace Android
} // namespace CCL

using namespace CCL;
using namespace CCL::Android;

//************************************************************************************************
// VulkanSurfaceView
//************************************************************************************************

DEFINE_CLASS_ABSTRACT_HIDDEN (VulkanSurfaceView, Object)

//////////////////////////////////////////////////////////////////////////////////////////////////

VulkanSurfaceView::VulkanSurfaceView (Window* parent, VulkanWindowRenderTarget* renderTarget)
: parent (parent),
  renderTarget (renderTarget),
  nativeWindow (nullptr)
{
	JniAccessor jni;
	androidView.assign (jni, jni.newObject (RenderSurfaceView, RenderSurfaceView.construct, asIntPtr (), FrameworkActivity::getCurrentActivity ()->getJObject ()));

	setSize (parent->getSize ());

	// add surface view to the window's content view
	if(FrameworkView* frameworkView = AndroidWindow::cast (parent)->getFrameworkView ())
		ViewGroup.addView (*frameworkView, androidView);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

VulkanSurfaceView::~VulkanSurfaceView ()
{
	if(!androidView)
		return;

	cancelSignals ();

	JniAccessor jni;

	// remove view
	if(FrameworkView* frameworkView = AndroidWindow::cast (parent)->getFrameworkView ())
		ViewGroup.removeView (*frameworkView, androidView);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

void VulkanSurfaceView::setSize (RectRef size)
{
	Rect zoomedSize = Rect (size).zoom (renderTarget->getScaleFactor ());

	JniAccessor jni;
	RenderSurfaceView.layout (androidView, zoomedSize.left, zoomedSize.top, zoomedSize.right, zoomedSize.bottom);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

void VulkanSurfaceView::startRendering ()
{
	JniAccessor jni;
	RenderSurfaceView.startRendering (androidView);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

void VulkanSurfaceView::stopRendering ()
{
	JniAccessor jni;
	RenderSurfaceView.stopRendering (androidView);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

void VulkanSurfaceView::notify (ISubject* subject, MessageRef msg)
{
	if(msg == "render")
		onRender ();
	else if(msg == "surfaceCreated")
		onSurfaceCreated (msg[0]);
	else if(msg == "surfaceDestroyed")
		onSurfaceDestroyed ();
	else if(msg == "surfaceResized")
		onSurfaceResized (msg[0], msg[1]);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

void VulkanSurfaceView::onSurfaceCreated (VariantRef surface)
{
	JniAccessor jni;
	if(!RenderSurfaceView.isAlive (androidView))
		return;

	nativeWindow = (ANativeWindow*) surface.asIntPointer ();

	renderTarget->onSurfaceCreated (nativeWindow);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

void VulkanSurfaceView::onSurfaceDestroyed ()
{
	if(!nativeWindow)
		return;

	renderTarget->onSurfaceDestroyed ();

	ANativeWindow_release (nativeWindow);

	nativeWindow = nullptr;
}

//////////////////////////////////////////////////////////////////////////////////////////////////

void VulkanSurfaceView::onSurfaceResized (int width, int height)
{
	if(!nativeWindow)
		return;

	renderTarget->onSurfaceResized (width, height);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

void VulkanSurfaceView::onRender ()
{
	if(!nativeWindow)
		return;

	renderTarget->onRender ();
}

//////////////////////////////////////////////////////////////////////////////////////////////////

DECLARE_JNI_CLASS_METHOD_CCLGUI (void, VulkanSurfaceView, onSurfaceCreatedNative, JniIntPtr nativeViewPtr, jobject surface)
{
	VulkanSurfaceView* nativeView = VulkanSurfaceView::fromIntPtr (nativeViewPtr);
	ASSERT (nativeView)
	if(!nativeView)
		return;

	JniAccessor jni;
	ANativeWindow* nativeWindow = ANativeWindow_fromSurface (jni, surface);

	(NEW Message ("surfaceCreated", UIntPtr (nativeWindow)))->postBlocking (nativeView);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

DECLARE_JNI_CLASS_METHOD_CCLGUI (void, VulkanSurfaceView, onSurfaceDestroyedNative, JniIntPtr nativeViewPtr)
{
	VulkanSurfaceView* nativeView = VulkanSurfaceView::fromIntPtr (nativeViewPtr);
	ASSERT (nativeView)
	if(!nativeView)
		return;

	(NEW Message ("surfaceDestroyed"))->postBlocking (nativeView);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

DECLARE_JNI_CLASS_METHOD_CCLGUI (void, VulkanSurfaceView, onSurfaceResizedNative, JniIntPtr nativeViewPtr, int width, int height)
{
	VulkanSurfaceView* nativeView = VulkanSurfaceView::fromIntPtr (nativeViewPtr);
	ASSERT (nativeView)
	if(!nativeView)
		return;

	(NEW Message ("surfaceResized", width, height))->post (nativeView);
}

//////////////////////////////////////////////////////////////////////////////////////////////////

DECLARE_JNI_CLASS_METHOD_CCLGUI (void, VulkanSurfaceView, onRenderNative, JniIntPtr nativeViewPtr)
{
	VulkanSurfaceView* nativeView = VulkanSurfaceView::fromIntPtr (nativeViewPtr);
	ASSERT (nativeView)
	if(!nativeView)
		return;

	// Post the render message only when the application is active as not returning to the Java
	// side in an inactive application blocks the main thread so the message is never processed.
	if(GUI.isApplicationActive ())
		(NEW Message ("render"))->postBlocking (nativeView);
}

File: ccl/platform/android/vulkan/java/VulkanSurfaceView.java
//************************************************************************************************
//
// This file is part of Crystal Class Library (R)
// Copyright (c) 2025 CCL Software Licensing GmbH.
// All Rights Reserved.
//
// Licensed for use under either:
//  1. a Commercial License provided by CCL Software Licensing GmbH, or
//  2. GNU Affero General Public License v3.0 (AGPLv3).
// 
// You must choose and comply with one of the above licensing options.
// For more information, please visit ccl.dev.
//
// Filename    : ccl/platform/android/vulkan/java/VulkanSurfaceView.java
// Description : Vulkan view implemented as a SurfaceView
//
//************************************************************************************************

package dev.ccl.cclgui;

import android.content.Context;
import android.graphics.PixelFormat;
import android.view.*;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;

//************************************************************************************************
// VulkanRenderThread
//************************************************************************************************

class VulkanRenderThread extends Thread
{
    protected VulkanSurfaceView surfaceView;

    private boolean isDone = false;
    private boolean hasSurface = false;
    private boolean waitingForSurface = true;
	private boolean paused = true;

    private int width = 0;
    private int height = 0;

    public VulkanRenderThread (VulkanSurfaceView surfaceView)
	{
        super ();

        this.surfaceView = surfaceView;
    }

    @Override
    public void run ()
	{
		setName ("VulkanRenderThread " + getId ());

		// loop until asked to quit
		try
		{
			while(!isDone)
			{
				// update asynchronous state (window size)
				int w, h;
				synchronized(this)
				{
					while(needToWait ())
					{
						if(!hasSurface)
						{
							if(!waitingForSurface)
							{
								waitingForSurface = true;
								notify ();
							}
						}
						wait ();
					}

					if(isDone)
						break;

					w = width;
					h = height;

					if(hasSurface && waitingForSurface)
						waitingForSurface = false;
				}
				
				// draw a frame
				if(w > 0 && h > 0 && !paused)
					surfaceView.onRender ();
			}
		}
		catch(InterruptedException e)
		{}
    }

    private boolean needToWait ()
	{
        if(isDone)
            return false;

        if(!hasSurface || paused)
            return true;

        if(width > 0 && height > 0)
            return false;

        return true;
    }

    public void surfaceCreated ()
	{
        synchronized(this)
		{
            hasSurface = true;
            notify ();
        }
    }

    public void surfaceDestroyed ()
	{
        synchronized(this)
		{
            hasSurface = false;
            notify ();

            while(!waitingForSurface && isAlive ())
			{
                try
				{
                    wait ();
                }
				catch(InterruptedException e)
				{
                    Thread.currentThread ().interrupt ();
                }
            }
        }
    }

    public void surfaceResized (int w, int h)
	{
        synchronized(this)
		{
            width = w;
            height = h;
            notify ();
        }
    }

	public void pauseRendering ()
	{
	    synchronized(this)
		{
            paused = true;
        }
	}
	
	public void resumeRendering ()
	{
	    synchronized(this)
		{
            paused = false;
            notify ();
        }
	}

    public void requestExitAndWait ()
	{
        synchronized(this)
		{
            isDone = true;
            notify ();
        }

        try
		{
            join ();
        }
		catch(InterruptedException e)
		{
            Thread.currentThread ().interrupt ();
        }
    }

	public boolean hasSurface ()
	{
        synchronized(this)
		{
            return hasSurface;
        }
	}
}

//************************************************************************************************
// VulkanSurfaceView
//************************************************************************************************

@Keep
public class VulkanSurfaceView extends SurfaceView
							   implements SurfaceHolder.Callback2
{
	private final long nativeViewPtr;

	private VulkanRenderThread renderThread = null;
	private boolean isAttached = false;
	private boolean startRequested = false;

	private int width = 0;
	private int height = 0;

	public VulkanSurfaceView (long nativeViewPtr, Context context)
	{
		super (context);

		this.nativeViewPtr = nativeViewPtr;

		setZOrderOnTop (true);
		setAlpha (1.f);

		getHolder ().setFormat (PixelFormat.TRANSPARENT);
		getHolder ().addCallback (this);
	}

	@Override
    protected void onAttachedToWindow ()
	{
		super.onAttachedToWindow ();

		if(!isAttached)
		{
			// start render thread
			renderThread = new VulkanRenderThread (this);
			renderThread.start ();

			if(startRequested)
				startRendering ();

			isAttached = true;
		}
    }
	
    @Override
    protected void onDetachedFromWindow ()
	{
        if(renderThread != null)
        	renderThread.requestExitAndWait ();

        isAttached = false;
        super.onDetachedFromWindow ();
    }
	
    public void startRendering ()
	{
		if(renderThread != null)
		{
			renderThread.resumeRendering ();

			startRequested = false;
		}
		else
			startRequested = true;
    }

    public void stopRendering ()
	{
		if(renderThread != null)
			renderThread.pauseRendering ();

		startRequested = false;
    }

    public boolean isAlive ()
	{
		if(renderThread == null)
			return false;

		return renderThread.hasSurface ();
    }

	// forward surface callbacks to render thread

	@Override
	public void surfaceCreated (@NonNull SurfaceHolder holder)
	{
		renderThread.surfaceCreated ();

		onSurfaceCreatedNative (nativeViewPtr, getHolder ().getSurface ());
		if(width > 0 && height > 0)
			onSurfaceResizedNative (nativeViewPtr, width, height);
	}

	@Override
	public void surfaceDestroyed (@NonNull SurfaceHolder holder)
	{
		renderThread.surfaceDestroyed ();

		onSurfaceDestroyedNative (nativeViewPtr);
	}

	@Override
	public void surfaceChanged (@NonNull SurfaceHolder holder, int format, int w, int h)
	{
		width = w;
		height = h;

		renderThread.surfaceResized (width, height);

		onSurfaceResizedNative (nativeViewPtr, width, height);
	}

	@Override
	public void surfaceRedrawNeeded (@NonNull SurfaceHolder holder)
	{}

	// forward render callback to native code

	public void onRender ()
	{
		onRenderNative (nativeViewPtr);
	}

	public native void onSurfaceCreatedNative (long nativeViewPtr, Surface surface);
	public native void onSurfaceDestroyedNative (long nativeViewPtr);
	public native void onSurfaceResizedNative (long nativeViewPtr, int width, int height);
	public native void onRenderNative (long nativeViewPtr);
}
